src/main/java/usi/si/seart/service/StaticCodeAnalysisServiceImpl.java

```diff
@@ -0,0 +1,109 @@
+        try (ClonedRepo clonedRepo = gitRepoClonerService.cloneRepo(new URL("https://github.com/" + repo.getName())).get()) {
+            // Runs cloc for gathering code metrics
+            String output = new TerminalExecution(clonedRepo.getPath(), "cloc --json --quiet .").start().getStdout().lines().collect(Collectors.joining("\n"));
+
+            // TODO: I think this can lead to side effects
+//            metrics = conversionService.convert(Pair.of(repo,g.fromJson(output, JsonObject.class)), HashSet.class);
+//            metrics = (new JsonObjectToRepoMetrics()).convert(g.fromJson(output, JsonObject.class)));
+            if (!persist) {
+                return new AsyncResult<>(convert(null, g.fromJson(output, JsonObject.class)));
+            }
+
+            // Converts the string output from 'cloc' into a set of code metrics.
+            metrics = convert(repo, g.fromJson(output, JsonObject.class));
+            repo.setMetrics(metrics);
+            repo.setCloned();
+            gitRepoRepository.save(repo);
+            log.info("Stored code metrics for repository '"+repo.getName()+"'");
+
+
+        } catch (InterruptedException | ExecutionException | TerminalExecutionException | MalformedURLException e) {
+            log.error("Could not compute code metrics for repository '"+repo.getName()+"'", e);
+            throw new StaticCodeAnalysisException(e);
+        }
+
+        return new AsyncResult<>(metrics);
+    }
+
+
+    public Future<HashSet<GitRepoMetric>> getCodeMetrics(@NotNull String repo_name, boolean persist) throws StaticCodeAnalysisException {
+        GitRepo repo = gitRepoRepository.findGitRepoByName(repo_name).orElse(null);
+        if (repo==null)
+            log.warn("Computing metrics for '"+repo_name+"' but did NOT find repo in the database.");
+
+        return getCodeMetrics(repo, persist);
+
+    }
+
+    private HashSet<GitRepoMetric> convert(@Nullable GitRepo repo, @NotNull JsonObject source) {
+        HashSet<GitRepoMetric> set = new HashSet<>();
+        source.entrySet().stream().filter((Map.Entry<String, JsonElement> entry) ->
+                !entry.getKey().equals("header") && !entry.getKey().equals("SUM")
+        ).forEach(entry -> {
+            JsonObject stat = entry.getValue().getAsJsonObject();
+            GitRepoMetric.GitRepoMetricBuilder builder = GitRepoMetric.builder();
+
+            MetricLanguage language = MetricLanguage.builder().language(entry.getKey()).build();
+            builder.language(language);
+            if (repo != null) {
+                builder.repo(repo);
+                builder.id(new GitRepoMetricKey(repo.getId(), language.getLanguage()));
+            }
+            builder.blankLines(stat.get("blank").getAsLong());
+            builder.commentLines(stat.get("comment").getAsLong());
+            builder.codeLines(stat.get("code").getAsLong());
+
+            set.add(builder.build());
+        });
+        return set;
```

Comment:
Rather than declaring the set, and then adding values to it manually in a `forEach`, use `map` in combination with `collect`. You can use `Collectiors.toSet()` as the argument for the latter if you do not plan on modifying the set later.